customers/services/profile.py

```diff
@@ -0,0 +1,128 @@
+import json
+from typing import Dict
+from uuid import UUID
+
+import requests
+from django.contrib.auth import get_user_model
+from django.http import HttpRequest
+
+from utils.relay import from_global_id, to_global_id
+
+from ..models import CustomerProfile
+
+User = get_user_model()
+
+PROFILE_API_URL = "PROFILE_API_URL"
+
+
+class ProfileService:
+    profile_token: str
+    api_url = "https://profiili-api.prod.kuva.hel.ninja/graphql/"
+    profile_list_threshold = 100
+
+    ALL_PROFILES_QUERY = """
+        query GetProfiles {
+            profiles(serviceType: BERTH) {
+                edges {
+                    node {
+                        id
+                        first_name: firstName
+                        last_name: lastName
+                        primary_email: primaryEmail {
+                            email
+                        }
+                    }
+                }
+            }
+        }
+    """
+
+    def __init__(self, request, profile_list_threshold=None, **kwargs):
+        if "config" in kwargs:
+            self.config = kwargs.get("config")
+
+        self.profile_token = self._get_city_profile_token(request)
+        self.profile_list_threshold = (
+            profile_list_threshold or self.profile_list_threshold
+        )
+        self.api_url = self.config.get(PROFILE_API_URL)
+
+    @staticmethod
+    def get_config_template():
+        return {
+            PROFILE_API_URL: str,
+        }
+
+    def _get_city_profile_token(self, request: HttpRequest, nullable: bool = False):
+        api_tokens = request.META.get("HTTP_API_TOKENS")
+        if not api_tokens:
+            raise Exception("Missing tokens")
+
+        api_tokens = json.loads(api_tokens)
+        profile_token = api_tokens.get("https://api.hel.fi/auth/helsinkiprofile")
+        if not nullable and not profile_token:
+            raise Exception("No token found in the request")
+
+        return profile_token
+
+    def get_all_profiles(self) -> Dict[str, CustomerProfile]:
+        query = self.ALL_PROFILES_QUERY
+
+        response = self.query(query)
+        edges = response.get("profiles", {}).get("edges", [])
+
+        # Parse the users received from the Profile Service
+        users = {}
+        for edge in edges:
+            user = self.parse_user_edge(edge)
+            users[user.id] = user
+
+        # Query all the profiles associated to them.
+        # After the loop to do it on a single query.
+        customer_profiles = CustomerProfile.objects.filter(id__in=users.keys())
+
+        # Associate the customer profiles with the fetched from the Profile Service
+        for profile in customer_profiles:
+            user = users[profile.id]
+            profile.user = user
+            users[profile.id] = profile
+
+        return users
+
+    def get_profile(self, id: UUID) -> CustomerProfile:
```

Comment:
Could this function just return the profile data from Hki profile API (id, firs namet, last name, email) instead of our CustomerProfile?

Reply:
(See the previous comment)